packages/navigator/src/browser/fuzzy-search.ts

```diff
@@ -0,0 +1,137 @@
+/*
+ * Copyright (C) 2018 TypeFox and others.
+ *
+ * Licensed under the Apache License, Version 2.0 (the 'License'); you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at http://www.apache.org/licenses/LICENSE-2.0
+ */
+
+import * as fuzzy from 'fuzzy';
+import { injectable } from 'inversify';
+
+/**
+ * Fuzzy searcher.
+ */
+export namespace FuzzySearch {
+
+    /**
+     * A range representing the match region.
+     */
+    export interface Range {
+
+        /**
+         * The zero based offset of the match region.
+         */
+        readonly offset: number;
+
+        /**
+         * The length of the match region.
+         */
+        readonly length: number;
+    }
+
+    /**
+     * A fuzzy search match.
+     */
+    export interface Match<T> {
+
+        /**
+         * The original item.
+         */
+        readonly item: T;
+
+        /**
+         * An array of ranges representing the match regions.
+         */
+        readonly ranges: ReadonlyArray<Range>;
+    }
+
+    /**
+     * The fuzzy search input.
+     */
+    export interface Input<T> {
+
+        /**
+         * The pattern to match.
+         */
+        readonly pattern: string;
+
+        /**
+         * The items to filter based on the `pattern`.
+         */
+        readonly items: ReadonlyArray<T>;
+
+        /**
+         * Function that extracts the string from the inputs which will be used to evaluate the fuzzy matching filter.
+         */
+        readonly transform: (item: T) => string;
+
+    }
+
+    /**
+     * The fuzzy search filter.
+     */
+    export const Search = Symbol('FuzzySearch.Search');
+    export interface Search {
```

Comment:
`FuzzySearch.Search` sounds verbose, is it because `FuzzySearch` collides with a const `FuzzySearch`? Can we do it somehow different that clients can inject `FuzzySearch`? for example, we can get rid of an interface and symbol and turn `FuzzySearchImpl` into `FuzzySearch`.  One still can subclass it and rebind.